funnel/views/participant.py

```diff
@@ -0,0 +1,164 @@
+def participants_json(profile, space):
+    return jsonp(participants=[participant_data(participant, space.id) for participant in space.participants])
+
+
+@app.route('/<space>/participants/new', methods=['GET', 'POST'], subdomain='<profile>')
+@lastuser.requires_login
+@load_models(
+    (Profile, {'name': 'profile'}, 'g.profile'),
+    ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
+    permission='new-participant')
+def new_participant(profile, space):
+    form = ParticipantForm()
+    form.events.query = space.events
+    if form.validate_on_submit():
+        participant = Participant(proposal_space=space)
+        form.populate_obj(participant)
+        try:
+            db.session.add(participant)
+            db.session.commit()
+        except IntegrityError:
+            db.session.rollback()
+            flash(_(u"This participant already exists."), 'info')
+        return redirect(space.url_for('admin'), code=303)
+    return render_form(form=form, title=_(u"New Participant"), submit=_(u"Add Participant"))
+
+
+@app.route('/<space>/participant/<participant_id>/edit', methods=['GET', 'POST'], subdomain='<profile>')
+@lastuser.requires_login
+@load_models(
+    (Profile, {'name': 'profile'}, 'g.profile'),
+    ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
+    (Participant, {'id': 'participant_id'}, 'participant'),
+    permission='participant-edit')
+def participant_edit(profile, space, participant):
+    form = ParticipantForm(obj=participant, model=Participant)
+    form.events.query = space.events
+    if form.validate_on_submit():
+        form.populate_obj(participant)
+        db.session.commit()
+        flash(_(u"Your changes have been saved"), 'info')
+        return redirect(space.url_for('admin'), code=303)
+    return render_form(form=form, title=_(u"Edit Participant"), submit=_(u"Save changes"))
+
+
+@app.route('/<space>/participant', subdomain='<profile>')
+@lastuser.requires_login
+@load_models(
+    (Profile, {'name': 'profile'}, 'g.profile'),
+    ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
+    permission='view')
+def participant(profile, space):
+    participant = Participant.query.filter_by(puk=request.args.get('puk')).first()
+    if not participant:
+        return jsonp(message=u"Not found", code=404)
+    elif participant.key == request.args.get('key'):
+        try:
+            contact_exchange = ContactExchange(user_id=g.user.id, participant_id=participant.id, proposal_space_id=space.id)
+            db.session.add(contact_exchange)
+            db.session.commit()
+        except IntegrityError:
+            app.logger.warning(u"Contact Exchange already present")
+            db.session.rollback()
+        return jsonp(participant=participant_data(participant, space.id, full=True))
+    else:
+        return jsonp(message=u"Unauthorized", code=401)
```

Comment:
I'll recommend using `jsonify` instead of `jsonp` as you don't want cross-site access to this endpoint, and maybe use `@render_with(json=True)` for the entire function. You can then `return {'message': "Unauthorized"}, 401`